sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TrackNotImplementedException.cs

```diff
@@ -50,12 +51,32 @@ protected override void Initialize(SonarAnalysisContext context)
                         return;
                     }
 
-                    if (c.SemanticModel.GetTypeInfo(throwStatement.Expression).Type.Is(KnownType.System_NotImplementedException))
+                    ReportDiagnostic(c, throwStatement.Expression, throwStatement);
+                },
+                SyntaxKind.ThrowStatement);
+
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var creationExpression = (ObjectCreationExpressionSyntax)c.Node;
+                    var throwExpression = creationExpression.Parent;
+
+                    if (throwExpression.Kind() != SyntaxKindEx.ThrowExpression)
                     {
-                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, throwStatement.GetLocation()));
+                        return;
                     }
+
+                    ReportDiagnostic(c, creationExpression, throwExpression);
                 },
-                SyntaxKind.ThrowStatement);
+                SyntaxKind.ObjectCreationExpression);
+        }
+
+        private static void ReportDiagnostic(SyntaxNodeAnalysisContext c, ExpressionSyntax expression, SyntaxNode syntax)
```

Comment:
I suggest to improve the names of the parameters:
- s/expression/newExceptionExpression
- s/syntax/throwExpression

(even if `throwExpression` can actually be a statement, this name is better than `syntax`)